### PR TITLE
Data and workflow demo for demand response with human in the loop study

### PR DESCRIPTION
**Pull request by BainanXia**
[Pull Request doc](https://breakthrough-energy.github.io/docs/user/git_guide.html#d-pull-request)

### Purpose
Put source data, profile data and demo notebooks to generate the demand flexibility profiles used in the human-in-the-loop study into repo for future references. **No production code.**

### What the code is doing
- A submodule `human_in_the_loop` is created in `prereise/gather/flexibilitydata/`
- `data` folder contains the processed source data of EnergyCoupon project
- `profiles` folder contains the demand flexibility profiles generated by cGAN based on the source data
- `demo` folder contains the workflow demonstration of the flexibility profile generation from the source data
- The attribution file is updated to include the EnergyCoupon source data

### Testing
The generated profiles are used in a Texas case study.

### Where to look
`prereise/gather/flexibilitydata/human_in_the_loop/`

### Usage Example/Visuals
Check the demo notebooks

### Time estimate
10 min to understand the context, more to look into the details

### CODE REVIEW

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+import numpy as np
+import pandas as pd
+
+# from sklearn.preprocessing import MinMaxScaler, StandardScaler
+from sklearn.preprocessing import StandardScaler
+from tslearn.clustering import TimeSeriesKMeans
+from tslearn.utils import to_time_series_dataset
+
+# import torch
+# import torch.nn as nn
```
Comment (lanesmith) [inline prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Can some of these commented out package imports be removed?

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+# import os
+# import sys
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+
+# from sklearn.preprocessing import MinMaxScaler, StandardScaler
+from sklearn.preprocessing import StandardScaler
+from tslearn.clustering import TimeSeriesKMeans
+from tslearn.utils import to_time_series_dataset
```
Comment (lanesmith) [inline prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Do we need to include `sklearn` and `tslearn` in the `requirements.txt` file?

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+        # for i in range(0, len(cluster_weather), ts_length):
+        #     temp_temp = cluster_weather['Temperature'].iloc[i:i+ts_length].copy()
+        #     wind_temp = cluster_weather['Wind Speed'].iloc[i:i+ts_length].copy()
+        #     hum_temp = cluster_weather['Relative Humidity'].iloc[i:i+ts_length].copy()
+        #     # print(temp.describe())label='humidity', label='wind speed'
+        #     plt.plot(temp_temp.values, c='red', alpha=alpha, lw=lw)
+        #     plt.plot(hum_temp.values, c='blue', alpha=alpha, lw=lw)
+        #     plt.plot(wind_temp.values, c='green', alpha=alpha, lw=lw)
+        #     plt.xlabel(f'Hours (total {ts_length/4} hours, 15-min interval)')
+        #     plt.title(f'Cluster {cluster_num} Weather for {title}')
```
Comment (lanesmith) [inline prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Can this commented-out section be removed?

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+# from sklearn.preprocessing import MinMaxScaler, StandardScaler
+from sklearn.preprocessing import StandardScaler
+from tslearn.clustering import TimeSeriesKMeans
+from tslearn.utils import to_time_series_dataset
+
+# import torch
+# import torch.nn as nn
+
+
+def scale_df(df):
```
Comment (lanesmith) [inline prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: I'm guessing we're not going to care about including docstrings at this point, right?

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+        for i in range(0, len(df.columns)):
+            temp = df.iloc[:, i].copy()
+            if sum(temp.iloc[-size:].values == arr) == size:
+                temp_df.insert(0, j, df.iloc[:, i])
+                j += 1
+
+    flex_label = temp_df.iloc[-3, :].copy()
+
+    if plot_profile:
+        # plt.plot(temp_df.iloc[:-3, :], c='red', lw=0.2, alpha=0.2)
```
Comment (lanesmith) [inline prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Can this be removed?

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+def get_gen_flex_label(generated, seq_num, arr, plot=True):
+    flex_label = []
+
+    for i in generated.columns:
+        flex_avg = generated[i].iloc[15:19].copy()
+        flex_avg.loc[flex_avg == 0] = np.NaN
+        flex_avg = flex_avg.mean()
+
+        other_avg = generated[i].loc[np.r_[0:14, 18:-1]].mean()
```
Comment (lanesmith) [inline prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Slices are endpoint-exclusive, correct? Between `flex_avg` and `other_avg`, are some points being excluded and some being double counted?

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 2 above)
Comment (lanesmith) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: It looks like one of the demo notebooks also requires `torch`.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/similar day summer label.ipynb [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+    "    \n",
+    "    flex_label = np.array([])\n",
+    "    temperature_label = np.array([])\n",
+    "    price_label = np.array([])\n",
+    "    \n",
+    "    for i in temp.columns:\n",
+    "        flex_avg = temp[i].iloc[15:20].copy()\n",
+    "        flex_avg = flex_avg.mean()\n",
+    "        \n",
+    "        other_avg = (temp[i].loc[0:14]).mean()\n",
```
Comment (lanesmith) [inline prereise/gather/flexibilitydata/human_in_the_loop/demo/similar day summer label.ipynb]: Are we forgetting to include certain time steps in the `other_avg`?

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/winter label.ipynb [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+    "            continue\n",
+    "        else:\n",
+    "            existing_days.append(i)\n",
+    "            \n",
+    "        new_winter_day = pd.to_datetime(summer2winter[pd.to_datetime(i)])\n",
+    "        \n",
+    "        flex_avg = temp[i].iloc[15:20].copy()\n",
+    "        flex_avg = flex_avg.mean()\n",
+    "        \n",
+    "        other_avg = np.concatenate((temp[i].loc[0:14], temp[i].loc[20:])).mean()\n",
```
Comment (lanesmith) [inline prereise/gather/flexibilitydata/human_in_the_loop/demo/winter label.ipynb]: It looks like index 14 isn't being accounted for in either average.

**Review by lanesmith — APPROVED** [review]
Overall, it seems fine to me. The logic seems to follow what we've been discussing for these last few months. It doesn't seem like there is anything obvious missing. I left some comments regarding removing commented-out code, adding packages to `requirements.txt`, and some questions I had about some of the averaging.

**Comment by lanesmith** [comment]
Is there still a plan to rename `functions.py` as `helpers.py`? Additionally, do we want to replace the spaces in the file names with underscores?

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 1 above)
Comment (BainanXia) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Idk, they weren't commented out before. I did that because the linter complains. I will keep them there just in case Sienna might want to revisit this in the repo sometime in the future.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 2 above)
Comment (BainanXia) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Yeah, I noticed that. I didn't add those dependencies into `requirements.txt` considering that it's not production code, i.e. we might refactor the code with standard doc strings added and redundant dependencies removed, even using other packages that already in the list to do these clustering work.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 4 above)
Comment (BainanXia) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Yep.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 3 above)
Comment (BainanXia) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Same comment above.

**Comment by BainanXia** [comment]
> Is there still a plan to rename `functions.py` as `helpers.py`? Additionally, do we want to replace the spaces in the file names with underscores?

Ah, good catch. I forgot it. Will do.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/similar day summer label.ipynb (hunk 7 above)
Comment (sienna-shi-github) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/similar day summer label.ipynb]: This was done on purpose, since the user might choose to continue curtailing the load after the 3-7pm range, so I did not include those periods when calculating the label. Although I realized my Winter and Summer label process is not consistent. `winter label.ipynb` has been changed to reflect this.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 6 above)
Comment (sienna-shi-github) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: That's right. This will be fixed.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 5 above)
Comment (sienna-shi-github) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Yes.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/winter label.ipynb (hunk 8 above)
Comment (sienna-shi-github) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/winter label.ipynb]: You're right, this has been fixed.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 6 above)
Comment (sienna-shi-github) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: In `get_gen_flex_label` function

    for i in generated.columns:
        flex_avg = generated[i].iloc[15:19].copy()
        flex_avg.loc[flex_avg == 0] = np.NaN
        flex_avg = flex_avg.mean()
        
        other_avg = generated[i].loc[np.r_[0:14, 20:-1]].mean()
        delta = other_avg - flex_avg 
        
        if delta >= 0.5:
            flex_label = np.append(flex_label, 2)
        elif delta < 0.5 and delta >= 0.2:
            flex_label = np.append(flex_label, 1)
        elif delta < 0.2:
            flex_label = np.append(flex_label, 0)

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/similar day summer label.ipynb (hunk 7 above)
Comment (sienna-shi-github) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/similar day summer label.ipynb]: In cell 6

    for i in temp.columns:
        flex_avg = temp[i].iloc[15:20].copy()
        flex_avg = flex_avg.mean()
        
        other_avg = (temp[i].iloc[0:15]).mean()
        
        delta = (other_avg - flex_avg)/other_avg

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/winter label.ipynb (hunk 8 above)
Comment (sienna-shi-github) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/winter label.ipynb]: In cell 7
    
    temp = pd.read_csv(os.path.join(winterFlexPath, file))
    dates = temp.columns
    flex_label = np.array([])
    temperature_label = np.array([])
    price_label = np.array([])
    existing_days = []
    
    for i in temp.columns:
        
        if pd.to_datetime(i) not in summer2winter:
            continue
        else:
            existing_days.append(i)
            
        new_winter_day = pd.to_datetime(summer2winter[pd.to_datetime(i)])
        
        flex_avg = temp[i].iloc[15:20].copy()
        flex_avg = flex_avg.mean()
        
        other_avg = (temp[i].iloc[0:15]).mean()

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 5 above)
Comment (BainanXia) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Done.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py (hunk 6 above)
Comment (BainanXia) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/functions.py]: Done.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/similar day summer label.ipynb (hunk 7 above)
Comment (BainanXia) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/similar day summer label.ipynb]: Done.

File: prereise/gather/flexibilitydata/human_in_the_loop/demo/winter label.ipynb (hunk 8 above)
Comment (BainanXia) [reply prereise/gather/flexibilitydata/human_in_the_loop/demo/winter label.ipynb]: Done.
